### ISSUE TITLE
Cancelling collection offers

### ISSUE THREAD

**Issue by kocken** [ProjectOpenSea/opensea-js#1083]
Hello! I'd like to cancel collection offers using the API or JS SDK. The only way I think this is possible is by using the cancelOrder function in the SDK. The problem is that this function requires the order in the OrderV2 format, which is seemingly hard to accomplish.

I've now collected data from the "[retrieve collection offers](https://docs.opensea.io/reference/retrieve-collection-offers)" endpoint, which I'd like to use with the cancelOrder SDK function to cancel my placed collection offer. For this I need to map the collection offer data to the data structure of the OrderV2 object. All parameters need to be mapped in order to initialize this object, so it's quite hard as it is so many parameters.

So far I've mapped all values, looks like these (some are knowingly null):
const order: OrderV2 = {
    orderHash: offer.order_hash,
    listingTime: offer.protocol_data.parameters.startTime,
    createdDate: offer.protocol_data.parameters.startTime.toString(),
    expirationTime: offer.protocol_data.parameters.endTime,
    closingDate: offer.protocol_data.parameters.endTime.toString(),
    cancelled: false,
    finalized: false,
    markedInvalid: false,
    clientSignature: offer.protocol_data.signature,
    maker: {
        address: offer.protocol_data.parameters.offerer,
        config : null,
        user: null,
        profileImgUrl: null
    },
    taker: null,
    currentPrice: null, // this might be obtainable from the offer
    protocolData: null, // this might be obtainable from the offer
    makerAssetBundle: null,
    makerFees: null,
    orderType: null,
    protocolAddress: null,
    side: 'bid',
    takerAssetBundle: null,
    takerFees: null
};

But when I try to cancel this order as:
`await openSeaSdk.cancelOrder({
    accountAddress: accountAddress,
    order: order
});`

I receive the following error: Error: invalid address (argument="address", value=null, code=INVALID_ARGUMENT, version=address/5.7.0)

I'm not sure which "address" property is being required here, is it the protocolAddress value that is required to be mapped?
Furthermore, do I need to actually fully map additional (all? 😅) OrderV2 values in order to use the cancelOrder function for cancelling collection offers?

**Comment by ryanio** [comment]
`cancelOrder` cancels an order on chain. the only info that needs to be populated in the OrderV2 is the `order.protocolAddress` and `order.protocolData.parameters`

**Comment by kocken** [comment]
> `cancelOrder` cancels an order on chain. the only info that needs to be populated in the OrderV2 is the `order.protocolAddress` and `order.protocolData.parameters`

Thanks for the answer. Am I right to assume that this would cost gas fees then, considering it's on chain? I'd like to cancel collection offers without paying gas, the same way as I can do using the Opensea web interface

**Comment by thenerdassassin** [comment]
Yes, this would be an onchain transaction which costs gas. Currently the API and SDK do not support gasless cancellations. It is an item we have in our backlog but want to make sure to design the API in a way which makes it clear to the clients that gasless cancellation is best effort and not immediate.

This distinction is fairly easy to do through a UI since you can add loading screens or show the order is "pending cancellation". Also, you can pop up warnings to the user that the order will not be immediately cancelled when using gasless cancellation.

This is harder with an API since Rest APIs are generally used for immediate CRUD operations and are not supportive of an action which could take up to 5 minutes to resolve. All this to say, it is feature we want to add but we want to make sure it's addition is done with the best customer experience.

Sorry for any convenience this may cause by not being available now.

**Comment by kocken** [comment]
That's fair. If the operation takes long time to complete, perhaps you could implement a POST endpoint for initiating the gasless cancellation, and then add a GET endpoint for polling the status of this operation. In any case, I understand that this can take some time to implement.

No problem, thanks for the information!

**Comment by thenerdassassin** [comment]
I think the most likely implementation is we allow API users to have gasless cancellation but make it clear in developer documentation that cancelling without gas is considered best effort and the order may be filled before cancellation is finalized. The only guaranteed cancellation is on the immutable ledger that is the blockchain.

**Comment by kocken** [comment]
Sure, sounds like a very good solution

**Comment by kocken** [comment]
I'm realizing I'm in great need of this, what would the ETA be ish for this? I'm considering if it's worth to try bot the web interface or try reverse-engineer the graphql API for this, if you'd allow this 😅 A regular OS SDK/API endpoint would be preferred but I might need to attempt a solution until this is supported

**Comment by karanpokar** [comment]
I am facing the same issue. 
```
const res = await openseaSDK
      ?.cancelOrder({
        order: order?.[0]?.protocolData?.parameters,
        accountAddress: order?.[0]?.protocolAddress,
      })
      .catch((err) => {
        console.log('Err', err)
      })
```

Still getting the same issue->>
`Err Error: invalid address (argument="address", value=undefined, code=INVALID_ARGUMENT, version=address/5.7.0)
    at Logger.makeError (index.js:238:1)
    at Logger.throwError (index.js:247:1)
    at Logger.throwArgumentError (index.js:250:1)
    at Module.getAddress (index.js:71:1)
    at isValidProtocol (utils.js:391:1)
    at OpenSeaSDK.<anonymous> (sdk.js:551:1)
    at Generator.next (<anonymous>)
    at sdk.js:27:1
    at new Promise (<anonymous>)
    at push../node_modules/opensea-js/lib/sdk.js.__awaiter (sdk.js:9:1)`

**Comment by karanpokar** [comment]
i have changed this to this and still the same error

` const res = await openseaSDK
      ?.cancelOrder({
        order: {
          ...order?.[0]?.protocolData?.parameters,
          ...{ address: order?.[0]?.protocolAddress },
        },
        accountAddress: address,
      })`

**Comment by ryanio** [comment]
Keep the same shape for order don't do that spread operator stuff you are changing the expected object format. Just make sure the fields (protocolSata.parameters and protocolAddress) are not null on OrderV2.

**Comment by thenerdassassin** [comment]
> I'm realizing I'm in great need of this, what would the ETA be ish for this? I'm considering if it's worth to try bot the web interface or try reverse-engineer the graphql API for this, if you'd allow this 😅 A regular OS SDK/API endpoint would be preferred but I might need to attempt a solution until this is supported

Do not bot the web interface or reverse-engineer the graphql API. Both of these would violate the Developer Terms of Service which was agreed to when signing up for API Keys during Developer Onboarding and can lead to actions being take against your account including revoking your current keys.

**Comment by kocken** [comment]
> 

Fair, noted! Do you have any other suggestions that I might be able to use until this is supported, other than the onchain gas-transactions? I'm realizing that I might not be able to continue trading without this support

**Comment by thenerdassassin** [comment]
you can use on-chain transaction on the Seaport contract to increment the counter which will cancel all orders in a single transaction for really cheap. There is not another way to programmatically cancel right now.

However, hearing the needs of customers like this helps us to prioritize adding this API/SDK call over other features in the backlog.

**Comment by kocken** [comment]
Very interesting, what amount would I assign to the counter to cancel all placed collection offers? If only one one call is required to cancel all collection offers, then this could be a great solution.

That's great, happy to hear!

**Comment by thenerdassassin** [comment]
@kocken You can ask the Seaport contract, via etherscan or programmatically, what the current counter is for your wallet address and just set the counter to counter+1

https://etherscan.io/address/0x00000000000000adc04c56bf30ac9d3c0aaf14dc#readContract#F2

**Comment by Segfaultd** [comment]
> @kocken You can ask the Seaport contract, via etherscan or programmatically, what the current counter is for your wallet address and just set the counter to counter+1
> 
> https://etherscan.io/address/0x00000000000000adc04c56bf30ac9d3c0aaf14dc#readContract#F2

This should be absolutely documented somewhere on the OpenSea documentation !!
On Polygon, I wasn't able to use the builtin feature of the webapp to cancel all my orders. So I started doing it programatically but for 2500+ listings it's expensive and super super long.

Incrementing counter saved my day.

Please document it, or allow to use webapp mass cancel with Polygon :-)

**Comment by aviexk** [comment]
+1 on the need for gassless cancellations through the sdk - def need this to make sure my bids remain competitive, without which I am shooting in the dark

**Comment by ryanio** [comment]
Thanks for your patience, this is now available in opensea-js [v7.1.8](https://github.com/ProjectOpenSea/opensea-js/releases/tag/v7.1.8)
